writer.go

```diff
@@ -335,6 +338,55 @@ func NewMediaPlaylist(winsize uint, capacity uint) (*MediaPlaylist, error) {
 	return p, nil
 }
 
+// InsertSegments allows the insertion of one or multiple MediaSegments into the MediaPlaylist.
+// seqID is the sequence ID at which the new segments should be inserted to
+func (p *MediaPlaylist) InsertSegments(segments []MediaSegment, seqID uint64) error {
+	if len(segments) == 0 {
+		return ErrPlaylistEmpty
+	}
+
+	// Create copies of the segments to be inserted
+	localSegments := make([]*MediaSegment, len(segments))
+	for i, segment := range segments {
+		copySegment := segment // Create a shallow copy of the segment
+		localSegments[i] = &copySegment
+	}
+
+	adjustment := uint(len(localSegments))
+	p.count += adjustment
+	p.tail += adjustment
+
+	// Determine the index where the new segments should be inserted
+	var insertIndex int
+	// Allow passing in 0 for prerolls
+	if seqID == 0 {
+		insertIndex = 0
+	} else {
+		// Find the index where the new segments should be inserted
+		for i, segment := range p.Segments {
+			if segment.SeqId == seqID {
+				insertIndex = i + 1
+				break
+			}
+		}
+	}
+
+	// Insert the new segments into the playlist
+	p.Segments = append(p.Segments[:insertIndex], append(localSegments, p.Segments[insertIndex:]...)...)
+
+	for i := insertIndex; i < len(p.Segments[:insertIndex])+len(localSegments); i++ {
+		p.Segments[i].SeqId += uint64(len(p.Segments[:insertIndex]))
+	}
+
+	// Adjust the sequence IDs of the following segments
+	for i := insertIndex + len(localSegments); i < len(p.Segments); i++ {
+		if p.Segments[i] != nil {
```

Comment:
It's possible to have a `nil` segment here?

Reply:
Yes it is, if the capacity is larger than the amount of segments

Reply:
@KernelPryanic Whilst I was testing the other PR in SSAI, I had managed to get 1024 segments, majority with nil values when `decode()` is called:
`media, err = NewMediaPlaylist(8, 1024) // Winsize for VoD will become 0, capacity auto extends`

Reply:
When I didn't check for `nil` values, I got panic

Reply:
Ah, I see, so segments [can](https://github.com/jwplayer/m3u8/blob/32c68449b86a76903611f65fd52ae693ced6397c/structure.go#L111) actually be `nil`, another question is why, but I guess it will take time to understand why it's done this way. Also yeah, the "capacity" here [acts as an actual size](https://github.com/jwplayer/m3u8/blob/32c68449b86a76903611f65fd52ae693ced6397c/reader.go#L196) of the slice and [managed separately](https://github.com/jwplayer/m3u8/blob/32c68449b86a76903611f65fd52ae693ced6397c/reader.go#L506-L507), so you may also need to update this variable after the insert.

Reply:
Also the question is why there's a different internal "capacity" is used which is actually a slice size, instead of relying on the Go slice capacity 😄